src/roots.jl

```diff
@@ -1,15 +1,77 @@
 
 import IntervalArithmetic: diam, isinterior
+import Base: start, next, done, copy, eltype, iteratorsize
 
-export branch_and_prune, Bisection, Newton
+export branch_and_prune, Bisection, Newton, RootSearch
+export start, next, done, copy, step!, eltype, iteratorsize
 
 diam(x::Root) = diam(x.interval)
 
 Base.size(x::Interval) = (1,)
 
 isinterior{N}(X::IntervalBox{N}, Y::IntervalBox{N}) = all(isinterior.(X, Y))
 
+struct RootSearchState{T <: Union{Interval,IntervalBox}}
+    working::Vector{T}
+    outputs::Vector{Root{T}}
+end
+RootSearchState{T<:Union{Interval,IntervalBox}}(region::T) =
+    RootSearchState([region], Root{T}[])
+
+copy(state::RootSearchState) = RootSearchState(copy(working), copy(outputs))
 
+"""
+    RootSearch{R <: Union{Interval,IntervalBox}, S <: Contractor, T <: Real}
+
+Type implementing the `Base.Iterator` interface to the branch and prune routine.
+Returns the `RootSearchState` at each iteration. Note: Each iteration mutates
+the `RootSearchState`. Use `copy(state::RootSearchState)` to create an
+independent instance if necessary.
+"""
+struct RootSearch{R <: Union{Interval,IntervalBox}, S <: Contractor, T <: Real}
+    region::R
+    contractor::S
+    tolerance::T
+end
+
+eltype{R}(::Type{RootSearch{R}}) = RootSearchState{R}
+iteratorsize(::Type{RootSearch}) = SizeUnknown()
+
+function start(iter::RootSearch)
+    state = RootSearchState(iter.region)
+    sizehint!(state.outputs, 100)
+    sizehint!(state.working, 1000)
+    return state
+end
+
+"""
+    step!(state::RootSearchState, contractor, tolerance)
+
+Progress `state` by treating one of its `working` regions. Note: `state.working`
+is always modified. If a root is found, it is added to `state.outputs`.
+"""
+function step!(state::RootSearchState, contractor, tolerance)
+    X = pop!(state.working)
+    status, output = contractor(X, tolerance)
+    if status == :empty
+        return nothing
+    elseif status == :unique
+        push!(state.outputs, Root(output, :unique))
+    elseif diam(output) < tolerance
+        push!(state.outputs, Root(output, :unknown))
+    else # branch
+        X1, X2 = bisect(X)
+        push!(state.working, X1, X2)
+    end
+    return nothing
+end
+
+function next(iter::RootSearch, state::RootSearchState)
+    step!(state, iter.contractor, iter.tolerance)
+    return state, state
```

Comment:
It seems strange to be returning `state` twice here?

Reply:
Yeah, it's a little awkward. The second return value must be `state` (dictated by the interface) and the first return value is what a `for i in iter` loop puts into `i`. Right now I think the most useful Iterator is one that allows inspection of the algorithms progress and therefore `i` should be the internal state.

One could imagine a *different* Iterator for which `next()` returns individual Roots as they are found but that will only be possible once the new [Nullable Iterator interface](https://github.com/JuliaLang/julia/issues/18823) is done. (Because we cannot decide if the iteration is done until we find (or dont find) the last root, so it is impossible to define a `done()` method.) 

(I should also point out that a lot of other Iterators use an Int as their state variable and modify the iterator object itself at every step which seems very inelegant and confusing to me.)